File: solutions/Leetcode_1354/leetcode_1354.cc
#include <numeric>
#include <queue>
#include <vector>

using namespace std;

class Solution1
{
public:
    bool isPossible(vector<int> target)
    {
        long total = accumulate(target.begin(), target.end(), (long)0);
        int max_value;
        priority_queue<int> q(target.begin(), target.end());
        while (true)
        {
            max_value = q.top();
            q.pop();
            total -= max_value;
            if (max_value == 1 || total == 1)
                return true;
            if (max_value <= total || total == 0 || max_value % total == 0)
                return false;
            max_value %= total;
            total += max_value;
            q.push(max_value);
        }
    }
};

File: solutions/Leetcode_1354/leetcode_1354.py
import heapq
from typing import List


class Solution1:
    def is_possible(self, target):
        total = sum(target)
        target = [-x for x in target]
        heapq.heapify(target)
        while True:
            max_value = -heapq.heappop(target)
            total -= max_value
            if max_value == 1 or total == 1:
                return True
            if max_value <= total or total == 0 or max_value % total == 0:
                return False
            max_value %= total
            total += max_value
            heapq.heappush(target, -max_value)


class SolutionMINE:
    # Time Limit Exceed
    def is_possible(self, target: List[int]) -> bool:
        target = [-x for x in target]
        heapq.heapify(target)
        while target[0] != -1:
            max_value = heapq.heappop(target)
            max_value -= sum(target)
            if max_value >= 0:
                return False
            heapq.heappush(target, max_value)
        return True


if __name__ == '__main__':
    target = [8, 5]
    ans = SolutionMINE().is_possible(target)
    print(ans)
